File: problem144.py
#iterative approach
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []
        stack=[(root, False)]
        result=[]
        while(stack):
            curr, visited=stack.pop()
            if visited:
                result.append(curr.val)
            else:
                if curr.right:
                    stack.append((curr.right, False))
                if curr.left:
                    stack.append((curr.left, False))
                stack.append((curr, True))
        return result
        
        

File: problem144.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public IList<int> PreorderTraversal(TreeNode root) {
        IList<int> result = new List<int>();
        Stack<TreeNode> stack = new Stack<TreeNode>();
        stack.Push(root);
        
        if(root==null)
            return result;
        
        while(stack.Count>0)
        {
            TreeNode current = stack.Pop();
            result.Add(current.val);
            if(current.right!=null)
                stack.Push(current.right);
            if(current.left!=null)
                stack.Push(current.left);
        }
        
        return result;
    }
}

//time: O(N)
//Space: O(N)